[#14311] Non Blocking Publisher API

wburns opened the pull request:
Fixes #14311

Fixes #14389
Fixes #14177
Fixes #14166
Fixes #14014
Fixes #13981
Fixes #13919
Fixes #13917
Fixes #13912
Fixes #13911
Fixes #13910
Fixes #13863
Fixes #13634
Fixes #13633
Fixes #13608
Fixes #13606
Fixes #13371
Fixes #13363

Required for #14016 to remove flaky test failures

wburns commented on core/src/main/java/org/infinispan/CachePublisher.java:
```diff
@@ -0,0 +1,222 @@
+    * @return
+    * @param <R>
+    */
+   <R> CompletionStage<Void> performRemotely(Function<? super Publisher<CacheEntry<K, V>>, ? extends Publisher<R>> transformer,
```
This method is the only method that requires a bit more of implementation specifics around it to allow the injection of the cache, which the `ClusterPublisherManager` doesn't currently support. But I think it would be useful as users today may use the `Stream.forEach` method.

jabolina commented on core/src/main/java/org/infinispan/CachePublisher.java:
```diff
@@ -0,0 +1,222 @@
+import org.infinispan.util.function.SerializableBiConsumer;
+import org.infinispan.util.function.SerializableBiFunction;
+import org.infinispan.util.function.SerializableFunction;
+import org.reactivestreams.Publisher;
```
Should we use the JDK one instead of exposing RxJava?

wburns replied:
This is the reactive streams Publisher btw, which is not RxJava specific. Also the client currently exposes org.reactivestreams.Publisher but none of the others. The "new" API is the one that would have support for JDK Publisher.

jabolina commented on multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java:
```diff
@@ -157,39 +150,57 @@ public CompletableFuture<Boolean> containsEntry(K key, V value) {
+      return sizeInternal().toCompletableFuture();
+   }
+
+   private CompletionStage<Void> removeInternal(Predicate<? super V> p) {
```
This requires the predicate to be marshallable, but that should be fine. Especially how this improves the operation to not pull all data.

wburns replied:
Good point. We could even check if it is marshallable even and then do it based on that, but probably isn't worth. I can just change it back to more similar to what we had before.

wburns replied:
Actually, I forgot! `computeIfPresent` already states it should be marshallable, so we are fine 😌

jabolina reviewed (commented):
That looks good. I really like the examples using the multimap.

wburns commented:
Updated and added all the needed impl code. I did remove the `performRemotely` method I had prior for now, I think we can skip it for the time being and let it cook later if we find we want it.

pruivo commented:
I only took a quick look at the new interface, and I would probably make the interface more like the `Stream` interface and remove some arguments from the methods

* `parallelReduction(boolean)`: if the default is sequential, I would only have a method `parallelReduction()`.
* `filterKeys(Set)`: doesn't naming the method  `withKeys(Set)` make more sense? A method name filter without a predicate doesn't seem right. No null allowed!
* Add a new method `filterKeys(Predicate)`. It may be useful, I guess.
* `filterSegments(IntSet)`: same as above, rename to `withSegment(IntSet)`. No null allowed!
* Add a new method `filterSegments(IntPredicate)`
* Replace `deliveryGuarantee(DeliveryGuarantee)` with `atMostOnce()`, `atLeastOnce()` and `exactlyOnce()` methods.

wburns commented:
> I only took a quick look at the new interface, and I would probably make the interface more like the `Stream` interface and remove some arguments from the methods
> 
> * `parallelReduction(boolean)`: if the default is sequential, I would only have a method `parallelReduction()`.
> * `filterKeys(Set)`: doesn't naming the method  `withKeys(Set)` make more sense? A method name filter without a predicate doesn't seem right. No null allowed!
> * Add a new method `filterKeys(Predicate)`. It may be useful, I guess.
> * `filterSegments(IntSet)`: same as above, rename to `withSegment(IntSet)`. No null allowed!
> * Add a new method `filterSegments(IntPredicate)`
> * Replace `deliveryGuarantee(DeliveryGuarantee)` with `atMostOnce()`, `atLeastOnce()` and `exactlyOnce()` methods.

I had debated some of these earlier, so thanks for the feedback!

* I will add `parallelReduction` and `sequentialReduction` methods like on Stream.
* `withKeys(Set)` sounds better. Only reason I allowed null is if a user wants to override an existing value. I could add a `withAllKeys()` method maybe, wdyt?
* This is already provided by the transformer as needed by the user
* `withSegments(IntSet)` sounds good. Also think I should add a `withAllSegments()` method to be complete.
* `filterSegments(IntPredicate)` would work, it would require additional impl changes though and not just for the API, sadly. I think we can do this, but I will create an issue to add this later.
* Sure the new methods sound good to me.

pruivo commented:
> > I only took a quick look at the new interface, and I would probably make the interface more like the `Stream` interface and remove some arguments from the methods
> > 
> > * `parallelReduction(boolean)`: if the default is sequential, I would only have a method `parallelReduction()`.
> > * `filterKeys(Set)`: doesn't naming the method  `withKeys(Set)` make more sense? A method name filter without a predicate doesn't seem right. No null allowed!
> > * Add a new method `filterKeys(Predicate)`. It may be useful, I guess.
> > * `filterSegments(IntSet)`: same as above, rename to `withSegment(IntSet)`. No null allowed!
> > * Add a new method `filterSegments(IntPredicate)`
> > * Replace `deliveryGuarantee(DeliveryGuarantee)` with `atMostOnce()`, `atLeastOnce()` and `exactlyOnce()` methods.
> 
> I had debated some of these earlier, so thanks for the feedback!
> 
> * I will add `parallelReduction` and `sequentialReduction` methods like on Stream.
> * `withKeys(Set)` sounds better. Only reason I allowed null is if a user wants to override an existing value. I could add a `withAllKeys()` method maybe, wdyt?
> * This is already provided by the transformer as needed by the user
> * `withSegments(IntSet)` sounds good. Also think I should add a `withAllSegments()` method to be complete.
> * `filterSegments(IntPredicate)` would work, it would require additional impl changes though and not just for the API, sadly. I think we can do this, but I will create an issue to add this later.
> * Sure the new methods sound good to me.

Sounds good to me. `filterSegments` can be added in a follow-up PR if it gives to much trouble or when someone requests it.

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:
```diff
@@ -2020,6 +2021,12 @@ public CacheStream<K> parallelStream() {
+   @Override
+   public CachePublisher<K, V> cachePublisher() {
+      // TODO: need to implement
+      return null;
```
Oops missed this one.

pruivo commented on core/src/main/java/org/infinispan/cache/impl/EncoderCache.java:
```diff
@@ -1018,6 +1032,237 @@ public B setValue(B value) {
       }
    }
 
+   private static class KeyFunctionEncoder<I, O> implements Function<Publisher<I>, O>, InjectableComponent {
```
The `EncoderCache` is already a large file. What do you think of moving these classes outside?

wburns replied:
Sure

pruivo commented on core/src/main/java/org/infinispan/reactive/publisher/impl/SegmentPublisherSupplier.java:
```diff
@@ -1,16 +1,18 @@
- * values was from the given segment, only that all have been notified prior.
+ * A {@link Publisher} that also allows receiving segment completion information if desired. If segment information
+ * is needed use {@link #publisherWithSegments()} which publishes the {@link Notification} instances that can either
+ * contain a value or a segment completion notification. In that latter case the notification the notification should be
```
```suggestion
 * contain a value or a segment completion notification. In that latter case the notification should be
```

wburns commented on core/src/main/java/org/infinispan/CachePublisher.java:
```diff
@@ -0,0 +1,226 @@
+
+   /**
+    * Allows callers to configure the publisher to only publish keys or values that map to the given keys in the provided
+    * Set. Defaults to null or "all" keys.
```
Yeah, I meant to remove that part, just missed on refactoring.

wburns commented on core/src/main/java/org/infinispan/CachePublisher.java:
```diff
@@ -0,0 +1,226 @@
+    * Same as {@link #entryReduction(Function, Function)} except that the
+    * Functions must also implement <code>Serializable</code>
+    * <p>
+    * The compiler will pick this overload for lambda parameters, making them <code>Serializable</code>
```
What are you expecting?

pruivo replied:
That too, but I meant the missing description of the lines below :sweat_smile:

wburns commented:
Updated, rebased, added simple cache and tests for various modes.

jabolina reviewed (commented):
I guess the RESP stuff is a missing rebase?

wburns commented:
Hrmm... yeah not sure how that RESP stuff snuck in 🤦

wburns commented:
Updated and rebased.

wburns commented on core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:
```diff
@@ -149,6 +160,12 @@ public void start() {
             .maxIdle(configuration.expiration().maxIdle()).build();
       componentRegistry.start();
+
+      if (Configurations.needSegments(configuration)) {
+         maxSegments = configuration.clustering().hash().numSegments();
+      } else {
+         maxSegments = 1;
+      }
```
Good point, removing. Forgot it can't have a store.

wburns commented on core/src/main/java/org/infinispan/cache/impl/CachePublisherImpl.java:
```diff
@@ -0,0 +1,189 @@
+
+   @Override
+   public <R> CompletionStage<R> keyReduction(Function<? super Publisher<K>, ? extends CompletionStage<R>> transformer, Function<? super Publisher<R>, ? extends CompletionStage<R>> finalizer) {
+      InvocationContext ctx = invocationContextFactory != null ? invocationContextFactory.createInvocationContext(false, InvocationContextFactory.UNBOUNDED) : null;
```
While I fully agree on the other methods as they are not found by the method that invoked it but instead at a later invocation. All methods that take a transformer or finalizer will throw a NPE by the invoking thread. But I can add it so it is more explicit in here as luckily this is now shared by all the caches.

wburns commented:
Updated.

pruivo reviewed (approved):
LGTM. @jabolina, feel free to merge this after your review.

jabolina commented:
Thanks, @wburns! This is great!!
